ibtk/src/lagrangian/FEDataManager.cpp

```diff
@@ -1077,9 +1112,207 @@ FEDataManager::prolongData(const int f_data_idx,
+            for (CellIterator<NDIM> b(box); b; b++)
+            {
+                const CellIndex<NDIM>& i_c = b();
+                if (num_intersections(i_c) == 0 && patch_box.contains(i_c))
+                {
+                    libMesh::Point p;
+                    for (unsigned int d = 0; d < NDIM; ++d)
+                    {
+                        p(d) = patch_x_lower[d] + patch_dx[d] * (static_cast<double>(i_c(d) - patch_lower[d]) + 0.5);
+                    }
+                    static const double TOL = sqrt(std::numeric_limits<double>::epsilon());
+                    const libMesh::Point ref_coords = FEInterface::inverse_map(dim, X_fe_type, elem, p, TOL, false);
+                    if (FEInterface::on_reference_element(ref_coords, elem->type(), TOL))
+                    {
+                        intersection_ref_coords.push_back(ref_coords);
+                        intersection_indices.push_back(i_c);
+                        num_intersections(i_c) += 1;
+                    }
+                }
+            }
+
+            // Restore the nodal coordinates.
+            for (unsigned int k = 0; k < n_node; ++k)
+            {
+                elem->point(k) = s_node_cache[k];
+            }
+
+            // If there are no intersection points, then continue on to the next
+            // element.
+            if (intersection_ref_coords.empty()) continue;
+
+            // Evaluate the Lagrangian quantity at the Eulerian grid point and
+            // update the data on the grid.
+            for (unsigned int i = 0; i < n_vars; ++i)
+            {
+                F_dof_map_cache.dof_indices(elem, F_dof_indices[i], i);
+            }
+            get_values_for_interpolation(F_node, *F_petsc_vec, F_local_soln, F_dof_indices);
+            if (is_density) get_values_for_interpolation(X_node, *X_petsc_vec, X_local_soln, X_dof_indices);
+            F_fe->reinit(elem, &intersection_ref_coords);
+            if (X_fe != F_fe) X_fe->reinit(elem, &intersection_ref_coords);
+            for (unsigned int qp = 0; qp < intersection_ref_coords.size(); ++qp)
+            {
+                const CellIndex<NDIM>& i_c = intersection_indices[qp];
+                typedef boost::multi_array_types::index_range range;
+                double F_qp = interpolate(qp, F_node[boost::indices[range(0, F_dof_indices[0].size())][0]], phi_F);
+                if (is_density)
+                {
+                    jacobian(dX_ds, qp, X_node, dphi_X);
+                    F_qp /= std::abs(dX_ds.det());
+                }
+                (*f_data)(i_c) += F_qp / static_cast<double>(num_intersections(i_c));
+            }
+        }
+    }
+
+    VecRestoreArray(F_local_vec, &F_local_soln);
+    VecGhostRestoreLocalForm(F_global_vec, &F_local_vec);
+
+    VecRestoreArray(X_local_vec, &X_local_soln);
+    VecGhostRestoreLocalForm(X_global_vec, &X_local_vec);
+
+    IBTK_TIMER_STOP(t_prolong_data_cell);
+    return;
+} // prolongData_cell
```

Comment:
This function is nearly identical to the face-centered version. I would very much prefer if we combined them and just had a few if statements to switch between face/cell algorithms inside the element loop.